comp/dogstatsd/listeners/udp_test.go

```diff
@@ -27,32 +31,98 @@ import (
 	return fxutil.Test[listenerDeps](t, fx.Options(
+		telemetryimpl.MockModule(),
 		config.MockModule(),
 		pidmapimpl.Module(),
 		fx.Replace(config.MockParams{Overrides: overrides}),
 	))
 }
 
-func newPacketPoolManagerUDP(cfg config.Component) *packets.PoolManager {
-	packetPoolUDP := packets.NewPool(cfg.GetInt("dogstatsd_buffer_size"))
+func newPacketPoolManagerUDP(cfg config.Component, packetsTelemetryStore *packets.TelemetryStore) *packets.PoolManager {
+	packetPoolUDP := packets.NewPool(cfg.GetInt("dogstatsd_buffer_size"), packetsTelemetryStore)
 	return packets.NewPoolManager(packetPoolUDP)
 }
 
 func TestNewUDPListener(t *testing.T) {
 	deps := fulfillDepsWithConfig(t, map[string]interface{}{"dogstatsd_port": "__random__"})
-	s, err := NewUDPListener(nil, newPacketPoolManagerUDP(deps.Config), deps.Config, nil)
+	telemetryStore := NewTelemetryStore(nil, deps.Telemetry)
+	packetsTelemetryStore := packets.NewTelemetryStore(nil, deps.Telemetry)
+	s, err := NewUDPListener(nil, newPacketPoolManagerUDP(deps.Config, packetsTelemetryStore), deps.Config, nil, telemetryStore, packetsTelemetryStore)
+
 	assert.NotNil(t, s)
 	assert.Nil(t, err)
 
 	s.Stop()
 }
 
+func TestUDPListenerTelemetry(t *testing.T) {
+	port, err := getAvailableUDPPort()
+	require.Nil(t, err)
+	cfg := map[string]interface{}{}
+	cfg["dogstatsd_port"] = port
+	cfg["dogstatsd_non_local_traffic"] = false
+
+	packetChannel := make(chan packets.Packets)
+	deps := fulfillDepsWithConfig(t, cfg)
+	telemetryStore := NewTelemetryStore(nil, deps.Telemetry)
+	packetsTelemetryStore := packets.NewTelemetryStore(nil, deps.Telemetry)
+	s, err := NewUDPListener(packetChannel, newPacketPoolManagerUDP(deps.Config, packetsTelemetryStore), deps.Config, nil, telemetryStore, packetsTelemetryStore)
+	require.NotNil(t, s)
+
+	assert.Nil(t, err)
+
+	s.Listen()
+	defer s.Stop()
+
+	fmt.Println(fmt.Sprintf("127.0.0.1:%d", port))
+	conn, err := net.Dial("udp", fmt.Sprintf("127.0.0.1:%d", port))
+	defer conn.Close()
+	assert.Nil(t, err)
+	n, err := conn.Write([]byte("hello world"))
+	fmt.Println(n)
+	assert.Nil(t, err)
+
+	select {
+	case pkts := <-packetChannel:
+		packet := pkts[0]
+		assert.NotNil(t, packet)
+
+		telemetryMock, ok := deps.Telemetry.(telemetry.Mock)
+		assert.True(t, ok)
+
+		registry := telemetryMock.GetRegistry()
+		var packetsMetric []*dto.Metric
+		var bytesCountMetric []*dto.Metric
+		metricsFamily, err := registry.Gather()
+		assert.Nil(t, err)
+
+		for _, metric := range metricsFamily {
+			if metric.GetName() == "dogstatsd__udp_packets" {
+				packetsMetric = metric.GetMetric()
+			}
+			if metric.GetName() == "dogstatsd__udp_packets_bytes" {
+				bytesCountMetric = metric.GetMetric()
+			}
+		}
+
+		assert.NotNil(t, packetsMetric)
+		assert.NotNil(t, bytesCountMetric)
+
+		assert.Equal(t, float64(1), packetsMetric[0].GetCounter().GetValue())
+		assert.Equal(t, float64(11), bytesCountMetric[0].GetCounter().GetValue())
```

Comment:
This pattern seems to repeat quite a few times. Would it make sense to refactor it into a helper function?

Reply:
I do agree with that. I was mostly focusing in adding tests. I would prefer to leave the refactoring to a separate PR. I was thinking that we could improve the telemetry mock API to allow to fetch metrics in a easier way.